rpcserver.go

```diff
@@ -7768,81 +7759,62 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
 
 	rpcSpecifier := marshalAssetSpecifier(specifier)
 
-	resp, err := r.AddAssetBuyOrder(ctx, &rfqrpc.AddAssetBuyOrderRequest{
-		AssetSpecifier: &rpcSpecifier,
-		AssetMaxAmt:    req.AssetAmount,
-		Expiry:         uint64(expiryTimestamp.Unix()),
-		PeerPubKey:     peerPubKey[:],
-		TimeoutSeconds: uint32(
-			rfq.DefaultTimeout.Seconds(),
-		),
-	})
-	if err != nil {
-		return nil, fmt.Errorf("error adding buy order: %w", err)
+	type quoteWithInfo struct {
+		quote   *rfqrpc.PeerAcceptedBuyQuote
+		rate    *rfqmath.BigIntFixedPoint
+		channel rfq.ChannelWithSpecifier
 	}
 
-	var acceptedQuote *rfqrpc.PeerAcceptedBuyQuote
-	switch r := resp.Response.(type) {
-	case *rfqrpc.AddAssetBuyOrderResponse_AcceptedQuote:
-		acceptedQuote = r.AcceptedQuote
+	acquiredQuotes := make([]quoteWithInfo, 0)
 
-	case *rfqrpc.AddAssetBuyOrderResponse_InvalidQuote:
-		return nil, fmt.Errorf("peer %v sent back an invalid quote, "+
-			"status: %v", r.InvalidQuote.Peer,
-			r.InvalidQuote.Status.String())
+	// TODO(george): should we limit the max number of acquired quotes?
+	// Tapd nodes with hundreds or more of chans might face a bottleneck
+	// here. Could just acquire quotes that reach a certain threshold.
+	for peer, channels := range chanMap {
+		quote, err := r.AcquireBuyOrder(
+			ctx, &rpcSpecifier, req.AssetAmount, expiryTimestamp,
+			&peer,
+		)
+		if err != nil {
+			// TODO(george): should handle negotiation failures
+			// gracefully.
+			return nil, err
+		}
 
-	case *rfqrpc.AddAssetBuyOrderResponse_RejectedQuote:
-		return nil, fmt.Errorf("peer %v rejected the quote, code: %v, "+
-			"error message: %v", r.RejectedQuote.Peer,
-			r.RejectedQuote.ErrorCode, r.RejectedQuote.ErrorMessage)
+		rate, err := rfqrpc.UnmarshalFixedPoint(quote.AskAssetRate)
+		if err != nil {
+			return nil, err
+		}
 
-	default:
-		return nil, fmt.Errorf("unexpected response type: %T", r)
+		acquiredQuotes = append(acquiredQuotes, quoteWithInfo{
+			quote: quote,
+			rate:  rate,
+			// Since the channels are sorted, we now the value with
+			// the greatest remote balance is at index 0.
+			channel: channels[0],
```

Comment:
Shouldn't we retain _all_ the channels? As we can add many hop hints for a given edge to the invoice. Or is the idea that we want to start doing that in another change?

Reply:
> Something to keep in mind here is that when receiving shards our peer does non-strict forwarding on the LND level. The channel we pick here does not bind the actual link that will be used for the HTLC.

With the recent set of liquidity fixes, IIUC we're much stricter here now (being too lax was itself a bug). 

Looks like actually just the channel ID and pubkey are explicitly needed, and we fetch the policy using one of lnd's RPC calls.